### PR TITLE
Added support for synthetic animation/transition events

### PR DESCRIPTION
**Pull request by milesj**
After messing around trying to mimic and polyfill transition start/end/cancel events, I ended up just adding basic support for `transitionend` and some animations events.

I'm not sure how to test these since they are timed and the current jest/jsdom make it difficult, but they definitely work as I tested them in a browser alongside a few of my projects.

### CODE REVIEW

**Comment by jimfb** [comment]
@spicyj @syranide

**Comment by milesj** [comment]
Would be nice to have this for React 0.15 :+1:

**Comment by fatboy1215** [comment]
@milesj update the the pull request

**Comment by milesj** [comment]
Bump.

**Comment by milesj** [comment]
Anything? Even a simple note? :P

**Comment by jimfb** [comment]
@spicyj @syranide

**Comment by spicyj** [comment]
@jimfb Can't you review this?

**Comment by jimfb** [comment]
@spicyj No, not intelligently.  The event dispatch is still an inscrutable mystery to me, and you guys (I guess mainly you and @zpao) always have strong opinions about the way we name events/attributes.  There are some diffs in this area that I can review myself, but this probably isn't one of them.

To me, if I had to decide between having this change and not having this change in my copy of React, I'd accept and fix later if needed.  But you and @zpao like things to be more perfect before merging, which means this diff needs some more eyeballs.

You and @syranide have the best understanding of this code, which is why I sent it your way from the very beginning.

**Comment by milesj** [comment]
I can vouch that understanding the synthetic event system took quite some time.

File: src/renderers/dom/client/eventPlugins/SimpleEventPlugin.js [hunk 1]
```diff
@@ -39,6 +41,24 @@ var eventTypes = {
...
+      bubbled: keyOf({onAnimationEnd: true}),
+      captured: keyOf({onAnimationEndCapture: true}),
+    },
+  },
+  animationIteration: {
+    phasedRegistrationNames: {
+      bubbled: keyOf({onAnimationIteration: true}),
+      captured: keyOf({onAnimationIterationCapture: true}),
+    },
+  },
```
Comment (zpao) [inline src/renderers/dom/client/eventPlugins/SimpleEventPlugin.js]: Can you keep these ABC ordered please

File: src/renderers/dom/client/eventPlugins/SimpleEventPlugin.js [hunk 2]
```diff
@@ -24,6 +24,8 @@ var SyntheticDragEvent = require('SyntheticDragEvent');
 var SyntheticTouchEvent = require('SyntheticTouchEvent');
 var SyntheticUIEvent = require('SyntheticUIEvent');
 var SyntheticWheelEvent = require('SyntheticWheelEvent');
+var SyntheticAnimationEvent = require('SyntheticAnimationEvent');
+var SyntheticTransitionEvent = require('SyntheticTransitionEvent');
```
Comment (zpao) [inline src/renderers/dom/client/eventPlugins/SimpleEventPlugin.js]: And then also ABC order these requires as well

File: src/renderers/dom/client/ReactBrowserEventEmitter.js [hunk 3]
```diff
@@ -83,6 +84,9 @@ var reactTopListenersCounter = 0;
 // events so we don't include them here
 var topEventMapping = {
   topAbort: 'abort',
+  topAnimationStart: getVendorPrefixedEventName('animationstart', true),
+  topAnimationEnd: getVendorPrefixedEventName('animationend', true),
+  topAnimationIteration: getVendorPrefixedEventName('animationiteration', true),
```
Comment (zpao) [inline src/renderers/dom/client/ReactBrowserEventEmitter.js]: ditto

File: src/renderers/dom/client/utils/getVendorPrefixedEventName.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  if (prefixedEventNames[eventName]) {
+    return prefixedEventNames[eventName];
+
+  } else if (!vendorPrefixes[eventName]) {
+    return eventName;
+  }
+
+  var prefixMap = vendorPrefixes[eventName];
+  var validator = initialValidators[eventName];
+  var style = document.createElement('div').style;
```
Comment (zpao) [inline src/renderers/dom/client/utils/getVendorPrefixedEventName.js]: Can we cache this node? We don't need to make a new one every time

File: src/renderers/dom/client/utils/getVendorPrefixedEventName.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+  if (validator && !validator.checked) {
```
Comment (zpao) [inline src/renderers/dom/client/utils/getVendorPrefixedEventName.js]: Will we ever run these a second time? Shouldn't the results any of these get cached

File: src/renderers/dom/client/utils/getVendorPrefixedEventName.js [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+var prefixedEventNames = {};
+
+/**
+ * Attempts to determine the correct vendor prefixed event name.
+ *
+ * @param {string} eventName
+ * @param {boolean} returnDefault
+ * @returns {string}
+ */
+function getVendorPrefixedEventName(eventName, returnDefault) {
```
Comment (zpao) [inline src/renderers/dom/client/utils/getVendorPrefixedEventName.js]: I'm tempted to say we shouldn't have a `returnDefault` param and we should just make the return value nullable, then consumers could fall back to a default if they need to. Boolean args like this aren't awesome (and neither is returning an empty string)

File: src/renderers/dom/client/utils/getVendorPrefixedEventName.js (hunk 5 above)
Comment (milesj) [inline src/renderers/dom/client/utils/getVendorPrefixedEventName.js]: Correct, they will only run once. The `checked` property handles this -- but I could probably just execute the validators on load and forget these checks all together.

File: src/renderers/dom/client/utils/getVendorPrefixedEventName.js (hunk 6 above)
Comment (milesj) [inline src/renderers/dom/client/utils/getVendorPrefixedEventName.js]: I originally had it this way. I will change it back.

**Comment by milesj** [comment]
I have no idea what happened when I rebased off of master. I'm going to do a new pull request.

**Comment by milesj** [comment]
Recreated here: https://github.com/facebook/react/pull/6005

**Pull request closed without merging**